### PR TITLE
[2.3] Static Code Analysis for Components

### PR DESCRIPTION
**Pull request by kalessil**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | n/a |

Static Code Analysis with Php Inspections (EA Extended), no functional changes:
- [Filesystem] : nested ifs, not optimal ifs
- in_array() miss-uses
- class re-implements an interface of super(s)
- array_keys/array_values as foreach array - fixed cases where changes has sense

### CODE REVIEW

File: src/Symfony/Component/Filesystem/Filesystem.php [hunk 1]
```diff
@@ -339,9 +336,9 @@ public function makePathRelative($endPath, $startPath)
         $endPathRemainder = implode('/', array_slice($endPathArr, $index));
 
         // Construct $endPath from traversing to the common path, then to the remaining $endPath
-        $relativePath = $traverser.(strlen($endPathRemainder) > 0 ? $endPathRemainder.'/' : '');
+        $relativePath = $traverser.($endPathRemainder !== '' ? $endPathRemainder.'/' : '');
```
Comment (fabpot) [inline src/Symfony/Component/Filesystem/Filesystem.php]: We are using Yoda condition in Symfony, so it should be `'' !== $endPathRemainder` instead.

File: src/Symfony/Component/Filesystem/Filesystem.php [hunk 2]
```diff
@@ -339,9 +336,9 @@ public function makePathRelative($endPath, $startPath)
...
 
-        return (strlen($relativePath) === 0) ? './' : $relativePath;
+        return $relativePath === '' ? './' : $relativePath;
```
Comment (fabpot) [inline src/Symfony/Component/Filesystem/Filesystem.php]: same here

File: src/Symfony/Component/Filesystem/Filesystem.php (hunk 1 above)
Comment (kalessil) [inline src/Symfony/Component/Filesystem/Filesystem.php]: Will update, ths.

File: src/Symfony/Component/DependencyInjection/Dumper/GraphvizDumper.php [hunk 3]
```diff
@@ -173,7 +173,7 @@ private function findNodes()
         foreach ($container->getServiceIds() as $id) {
             $service = $container->get($id);
 
-            if (in_array($id, array_keys($container->getAliases()))) {
+            if (array_key_exists($id, $container->getAliases())) {
```
Comment (WouterJ) [inline src/Symfony/Component/DependencyInjection/Dumper/GraphvizDumper.php]: this does not have exactly the same behaviour (as `in_array` was not executed in strict mode). However, I don't think it'll break anything

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 4]
```diff
@@ -1549,7 +1549,7 @@ public function getLanguages()
         foreach (array_keys($languages) as $lang) {
             if (strstr($lang, '-')) {
                 $codes = explode('-', $lang);
-                if ($codes[0] == 'i') {
+                if ($codes[0] === 'i') {
```
Comment (WouterJ) [inline src/Symfony/Component/HttpFoundation/Request.php]: shouldn't these be yoda style too?

File: src/Symfony/Component/DependencyInjection/Dumper/GraphvizDumper.php (hunk 3 above)
Comment (kalessil) [inline src/Symfony/Component/DependencyInjection/Dumper/GraphvizDumper.php]: I double checked (was in doubt to push this or not), id/aliases are strings, so strict array_key_exists doesn't make any harm (just as you commented).

**Comment by xabbuh** [comment]
:+1:

File: src/Symfony/Component/CssSelector/Exception/ExpressionErrorException.php [hunk 5]
```diff
@@ -19,6 +19,6 @@
  *
  * @author Jean-François Simon <jeanfrancois.simon@sensiolabs.com>
  */
-class ExpressionErrorException extends ParseException implements ExceptionInterface
+class ExpressionErrorException extends ParseException
```
Comment (nicolas-grekas) [inline src/Symfony/Component/CssSelector/Exception/ExpressionErrorException.php]: What's the rational for removing the interface (here and below?)

Comment (xabbuh) [inline src/Symfony/Component/CssSelector/Exception/ExpressionErrorException.php]: `ParseException` already implements `ExceptionInterface`.

Comment (kalessil) [inline src/Symfony/Component/CssSelector/Exception/ExpressionErrorException.php]: These declarations is just dead code.

It's two components only which has this pattern, so I assume it's re-factoring artifacts (all cases I met pointing to re-factoring).

Comment (kalessil) [inline src/Symfony/Component/CssSelector/Exception/ExpressionErrorException.php]: All dropped implements related to "class re-implements an interface of super(s)"

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 6]
```diff
@@ -1549,7 +1549,7 @@ public function getLanguages()
         foreach (array_keys($languages) as $lang) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: May I suggest we remove all these foreach+array_keys also? They just lower perf and memory usage

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 7]
```diff
@@ -1558,7 +1558,7 @@ public function getLanguages()
                     }
                 } else {
                     for ($i = 0, $max = count($codes); $i < $max; $i++) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: Isn't this whole loop the same as calling implode?

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: OK never mind...

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 6 above)
Comment (kalessil) [inline src/Symfony/Component/HttpFoundation/Request.php]: It's not so much work (I recall 10-15 warnings from analyzer), but side-effect is unused variables.

I'm not sure, so it worth discussing with maintainers internally and mentioning me with final decision.

**Comment by aitboudad** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @kalessil.

**Pull request closed without merging**